include/swift/AST/DiagnosticEngine.h

```diff
@@ -935,6 +948,53 @@ namespace swift {
              "transactions must be closed LIFO");
     }
   };
+
+  /// Represents a diagnostic transaction which constructs a compound diagnostic
+  /// from any diagnostics emitted inside. A compound diagnostic consists of a
+  /// parent error, warning, or remark followed by a variable number of child
+  /// notes. The semantics are otherwise the same as a regular
+  /// DiagnosticTransaction.
+  class CompoundDiagnosticTransaction : DiagnosticTransaction {
+  public:
+    explicit CompoundDiagnosticTransaction(DiagnosticEngine &engine)
+        : DiagnosticTransaction(engine) {}
+
+    ~CompoundDiagnosticTransaction() {
+      if (IsOpen) {
+        commit();
+      }
+
+      if (Depth == 0) {
+        Engine.TransactionStrings.clear();
+        Engine.TransactionAllocator.Reset();
+      }
+    }
+
+    void commit() {
+      assert(PrevDiagnostics < Engine.TentativeDiagnostics.size() &&
+             "CompoundDiagnosticTransaction must contain at least one diag");
+
+      // The first diagnostic is assumed to be the parent. If this is not an
+      // error or warning, we'll assert later when trying to add children.
+      Diagnostic &parent = Engine.TentativeDiagnostics[PrevDiagnostics];
+
+      // Associate the children with the parent.
+      for (auto diag =
+               Engine.TentativeDiagnostics.begin() + PrevDiagnostics + 1;
+           diag != Engine.TentativeDiagnostics.end(); ++diag) {
+        diag->setIsChildNote(true);
+        parent.addChildNote(std::move(*diag));
+      }
+
+      // Erase the children, they'll be emitted alongside their parent.
+      Engine.TentativeDiagnostics.erase(Engine.TentativeDiagnostics.begin() +
+                                            PrevDiagnostics + 1,
+                                        Engine.TentativeDiagnostics.end());
```

Comment:
I think you _can't_ do this yet, because this is what's keeping the notes alive when you emit them.

Reply:
I think the parent owns the note as of line 986 here, and keeps it alive until  it's emitted when we iterate through the parent's child notes. I could be missing something though.

Reply:
Oops, of course. It's not the parent that owns the note; it's the DiagnosticEngine's ChildNotes array.

…but where does *that* get cleared?

Reply:
`ChildNotes` is part of `Diagnostic` though, not `DiagnosticEngine` (the diff on Github doesn't show this for some reason until you expand it). So it should just get cleared when the last transaction is popped off the stack and `TentativeDiagnostics` is emptied.

Reply:
Aargh, of course. I keep getting Diagnostic and DiagnosticInfo mixed up. Okay.